agentendpoint/config_task.go

```diff
@@ -0,0 +1,459 @@
+				}
+
+				result := pResult.GetResourceResults().GetResults()[i]
+				result.GetExecutionSteps()[validationStepIndex] = &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep{
+					Step: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep_Validation{
+						Validation: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation{
+							Outcome: outcome,
+						},
+					},
+					ErrMsg: errMsg,
+				}
+
+				// Continue to unmarshal all resources even if one has an error.
+			}
+		}
+	}
+
+	// TODO: check for resouce conflicts
+}
+
+func (c *configTask) checkState(ctx context.Context) {
+	// First populate check state results (for policies that do not have validation errors).
+	for i, a := range c.Task.GetConfig().GetConfigAssignments() {
+		asgnmnt := c.assignments[a.GetConfigAssignment()]
+		aResult := c.results.GetResults()[i]
+		for i, p := range a.GetPolicies() {
+			plcy := asgnmnt.policies[p.GetId()]
+			// Skip state check if this policy already has an error from a previous step.
+			if plcy.hasError {
+				continue
+			}
+			pResult := aResult.GetOsPolicyResults().GetResults()[i]
+			for i := range p.GetResources() {
+				result := pResult.GetResourceResults().GetResults()[i]
+				result.GetExecutionSteps()[checkStateStepIndex] = &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep{
+					Step: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep_CheckDesiredState{
+						CheckDesiredState: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_CheckDesiredState{
+							Outcome: agentendpointpb.ApplyConfigTaskOutput_ResourceResult_CheckDesiredState_SKIPPED,
+						},
+					},
+				}
+			}
+		}
+	}
+
+	// Actually run check state.
+	for i, a := range c.Task.GetConfig().GetConfigAssignments() {
+		ctx = clog.WithLabels(ctx, map[string]string{"config_assignment": a.GetConfigAssignment()})
+		asgnmnt := c.assignments[a.GetConfigAssignment()]
+		aResult := c.results.GetResults()[i]
+		for i, p := range a.GetPolicies() {
+			ctx = clog.WithLabels(ctx, map[string]string{"policy_id": p.GetId()})
+			plcy := asgnmnt.policies[p.GetId()]
+			// Skip state check if this policy already has an error from a previous step.
+			if plcy.hasError {
+				clog.Debugf(ctx, "Policy has error, skipping state check.")
+				continue
+			}
+			pResult := aResult.GetOsPolicyResults().GetResults()[i]
+			for i, r := range p.GetResources() {
+				ctx = clog.WithLabels(ctx, map[string]string{"resource_id": r.GetId()})
+				res := plcy.resources[r.GetId()]
+
+				outcome := agentendpointpb.ApplyConfigTaskOutput_ResourceResult_CheckDesiredState_IN_DESIRED_STATE
+				errMsg := ""
+				err := res.checkState(ctx)
+				if err != nil {
+					outcome = agentendpointpb.ApplyConfigTaskOutput_ResourceResult_CheckDesiredState_ERROR
+					plcy.hasError = true
+					errMsg = fmt.Sprintf("Error running desired state check: %v", err)
+					clog.Errorf(ctx, errMsg)
+				}
+
+				if res.enforcementNeeded {
```

Comment:
could this be renamed to `inDesiredState` as it seems to better align as a value set by checkState?

Reply:
Yes, I actually already renamed in in the follow up PR i was working on :)